### ISSUE TITLE
Feature request - expose im2col / col2im + make differentiable

### ISSUE THREAD

**Issue by apinto92** [pytorch/pytorch#1048]
Hi folks,

Currently in Theano I am using images2neibs to partition the image in smaller patches embedded in the architecture (an online step). However, in PyTorch, at first sight, no information regarding this feature is available.

Cheers

**Comment by soumith** [comment]
From what i understand, [unfold](http://pytorch.org/docs/tensors.html?highlight=unfold#torch.Tensor.unfold) is the equivalent torch operation, but we dont have automatic differentiation written for it yet.

**Comment by fmassa** [comment]
Might be easier to just wrap `im2col` and `col2im`.
I believe `unfold` was originally written to have efficient `im2col` just by playing with strides, but it turned out that the `contiguous` call needed by `gemm` made things slower than a dedicated function.
Furthermore, `unfold` does not accept padding / dilatation.
Are there people out there using `unfold` as is? If not, maybe we should replace it with an equivalent (in the spatial domain) `im2col`, which has it's backward pass implemented via `col2im`?

**Comment by csarofeen** [comment]
@soumith Does it make more sense to implement a im2col/col2im layer in THNN backend or promote im2col/col2im as tensor functions in TH and define them as differentiable tensor functions? I'm not entirely sure where the delineation is between TH and THNN in this case.

**Comment by soumith** [comment]
thnn is better for im2col, as it's a domain specific function

**Comment by soumith** [comment]
heard from Bryan at nvidia that they are sending a PR for this. No one else work on this, please.

**Comment by szagoruyko** [comment]
duplicate of https://github.com/pytorch/pytorch/issues/419

**Comment by fmassa** [comment]
Also note that part of the behavior of `im2col` can be obtained with `unfold` as in https://github.com/pytorch/pytorch/pull/1523 , but a dedicated implementation of `im2col` will definitely be more memory efficient.

**Comment by jbarker-nvidia** [comment]
@soumith I work for Bryan at NVIDIA.  I have an initial bare-bones implementation of im2col working, but before creating a PR I wanted to discuss design.  Initially I have added a dedicated Im2Col layer in nn.  I see there are also kernels for row2col and vol2col already though.  Would it make sense to have a single layer in nn that calls row2col/im2col/vol2col appropriately depending on the shape of the input tensor or would it be better to have independent layers for each of those cases?  If we go with the single layer solution, what should that layer be called?  Once we get that pinned down I should have a PR in a couple of days.

My working branch is here: https://github.com/jbarker-nvidia/pytorch/tree/im2col

**Comment by soumith** [comment]
@jbarker-nvidia i'm sorry i missed your comment and it's been 9 days (that's embarassing). I dont have good ideas on what to call it, but how about: `nn.functional.unfold`, or `nn.functional.d2col`? If you propose something, I'm fine with it too. and thanks for the upcoming PR!

**Comment by teracamo** [comment]
@jbarker-nvidia @soumith Hi! I would like to ask if there would a legit wrap for ```im2col``` and ```col2im```? I can live with using ```unfold``` as a substitute of ```im2col``` but what about ```col2im```? I also see szagoruyko wrapping in #419 but he doesn't seemed to have made a pull request, plus I think it doesn't support autograd.

**Comment by szagoruyko** [comment]
@teracamo there are both im2col and col2im with autograd support in https://github.com/szagoruyko/pyinn/blob/master/pyinn/im2col.py

**Comment by teracamo** [comment]
Tested and working! Awesome

**Comment by jbarker-nvidia** [comment]
@soumith (Finally) got round to opening a PR on this: https://github.com/pytorch/pytorch/pull/2580

My plan is to add col2im into the same im2col.cu and im2col.py files and then add separate files for row2col and vol2col.  Seems more straightforward than a monolithic unfolding layer - but happy to go down that route if preferred too.

**Comment by teracamo** [comment]
It would be really great to also have col2vol, very handy when working with medical images

**Comment by jbarker-nvidia** [comment]
PR [#2580](https://github.com/pytorch/pytorch/pull/2580) now has working im2col and col2im.  row2col and vol2col to follow in separate PR.

**Comment by zxytim** [comment]
@soumith Will this be scheduled on next release?

**Comment by soumith** [comment]
closed via https://github.com/pytorch/pytorch/pull/2580
